src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBarracks.java

```diff
@@ -0,0 +1,198 @@
+
+    /**
+     * Max hut level of the Barracks.
+     */
+    private static final int BARRACKS_HUT_MAX_LEVEL = 5;
+
+    /**
+     * Tower position offset.
+     */
+    private static final int TOWER_OFFSET = 8;
+
+    /**
+     * Constructor for the Barracks building.
+     *
+     * @param c Colony the building is in.
+     * @param l Location of the building.
+     */
+    public BuildingBarracks(final Colony c, final BlockPos l)
+    {
+        super(c, l);
+    }
+
+    @Override
+    public void onUpgradeComplete(final int newLevel)
+    {
+        final World world = getColony().getWorld();
+        if (world != null && newLevel < 5)
+        {
+            for (int i = 1; i <= newLevel; i++)
+            {
+                final Tuple<BlockPos, EnumFacing> tuple = getPositionAndFacingForLevel(i);
+
+                if(!(world.getBlockState(tuple.getFirst()).getBlock() instanceof BlockHutBarracksTower))
+                {
+                    world.setBlockState(tuple.getFirst(), ModBlocks.blockHutBarracksTower.getDefaultState().withProperty(BlockHutBarracksTower.FACING, tuple.getSecond()));
+                    getColony().addNewBuilding((TileEntityColonyBuilding) world.getTileEntity(tuple.getFirst()));
+
+                    final AbstractBuilding building = getColony().getBuilding(tuple.getFirst());
+                    if (building instanceof BuildingBarracksTower)
+                    {
+                        ((BuildingBarracksTower) building).addBarracks(getLocation());
+                    }
+                }
+            }
+        }
+        super.onUpgradeComplete(newLevel);
+    }
+
+    final Tuple<BlockPos, EnumFacing> getPositionAndFacingForLevel(final int level)
+    {
+        BlockPos position = getLocation();
+        int tempLevel = level;
+
+        if (isMirrored())
+        {
+            tempLevel += 1;
+        }
+
+        switch (getRotation())
+        {
+            case 1:
+                tempLevel += 3;
+                break;
+            case 2:
+                tempLevel += 1;
+                break;
+            case 3:
+                tempLevel += 2;
+                break;
+            default:
+                //do nothing
+        }
+
+        if (tempLevel > 4)
+        {
+            tempLevel -= 4;
+        }
+
+        EnumFacing facing = EnumFacing.NORTH;
+        switch (tempLevel)
+        {
+            case 1:
+                position = position.offset(EnumFacing.SOUTH, TOWER_OFFSET).offset(EnumFacing.WEST, TOWER_OFFSET);
+                break;
+            case 2:
+                position = position.offset(EnumFacing.NORTH, TOWER_OFFSET).offset(EnumFacing.EAST, TOWER_OFFSET);
+                facing = EnumFacing.SOUTH;
+                break;
+            case 3:
+                position = position.offset(EnumFacing.SOUTH, TOWER_OFFSET).offset(EnumFacing.EAST, TOWER_OFFSET);
+                facing = EnumFacing.WEST;
+                break;
+            case 4:
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 4 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)